jimfs/src/main/java/com/google/common/jimfs/SystemJimfsFileSystemProvider.java

```diff
@@ -0,0 +1,277 @@
+  @Deprecated
+  public SystemJimfsFileSystemProvider() {} // a public, no-arg constructor is required
+
+  @Override
+  public String getScheme() {
+    return URI_SCHEME;
+  }
+
+  @Override
+  public FileSystem newFileSystem(URI uri, Map<String, ?> env) throws IOException {
+    checkArgument(
+        uri.getScheme().equalsIgnoreCase(URI_SCHEME),
+        "uri (%s) scheme must be '%s'",
+        uri,
+        URI_SCHEME);
+    checkArgument(
+        isValidFileSystemUri(uri), "uri (%s) may not have a path, query or fragment", uri);
+    checkArgument(
+        env.get(FILE_SYSTEM_KEY) instanceof FileSystem,
+        "env map (%s) must contain key '%s' mapped to an instance of %s",
+        env,
+        FILE_SYSTEM_KEY,
+        FileSystem.class);
+
+    FileSystem fileSystem = (FileSystem) env.get(FILE_SYSTEM_KEY);
+    if (fileSystems.putIfAbsent(uri, fileSystem) != null) {
+      throw new FileSystemAlreadyExistsException(uri.toString());
+    }
+    return fileSystem;
+  }
+
+  @Override
+  public FileSystem getFileSystem(URI uri) {
+    FileSystem fileSystem = fileSystems.get(uri);
+    if (fileSystem == null) {
+      throw new FileSystemNotFoundException(uri.toString());
+    }
+    return fileSystem;
+  }
+
+  @Override
+  public Path getPath(URI uri) {
+    checkArgument(
+        URI_SCHEME.equalsIgnoreCase(uri.getScheme()),
+        "uri scheme does not match this provider: %s",
+        uri);
+
+    String path = uri.getPath();
+    checkArgument(!isNullOrEmpty(path), "uri must have a path: %s", uri);
+
+    return toPath(getFileSystem(toFileSystemUri(uri)), uri);
+  }
+
+  /**
+   * Returns whether or not the given URI is valid as a base file system URI. It must not have a
+   * path, query or fragment.
+   */
+  private static boolean isValidFileSystemUri(URI uri) {
+    // would like to just check null, but fragment appears to be the empty string when not present
+    return isNullOrEmpty(uri.getPath())
+        && isNullOrEmpty(uri.getQuery())
+        && isNullOrEmpty(uri.getFragment());
+  }
+
+  /** Returns the given URI with any path, query or fragment stripped off. */
+  private static URI toFileSystemUri(URI uri) {
+    try {
+      return new URI(
+          uri.getScheme(), uri.getUserInfo(), uri.getHost(), uri.getPort(), null, null, null);
+    } catch (URISyntaxException e) {
+      throw new AssertionError(e);
+    }
+  }
+
+  /** Invokes the {@code toPath(URI)} method on the given {@code FileSystem}. */
+  private static Path toPath(FileSystem fileSystem, URI uri) {
+    // We have to invoke this method by reflection because while the file system should be
+    // an instance of JimfsFileSystem, it may be loaded by a different class loader and as
+    // such appear to be a totally different class.
+    try {
+      Method toPath = fileSystem.getClass().getDeclaredMethod("toPath", URI.class);
+      return (Path) toPath.invoke(fileSystem, uri);
+    } catch (NoSuchMethodException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)